New Resource: `azurerm_compute_fleet`

sinbai opened the pull request:
<!--  All Submissions -->


## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review


## Description

<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. 

If this is a breaking change for users please detail how it cannot be avoided and why it should be made in a minor version of the provider -->

API: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/azurefleet/resource-manager/Microsoft.AzureFleet/stable/2024-11-01/azurefleet.json

Doc: https://learn.microsoft.com/en-us/azure/azure-compute-fleet/overview 

Note: TC test subscription account needs to register Microsoft.AzureFleet namespace

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”


<!-- You can erase any parts of this template below this point that are not applicable to your Pull Request. -->


## Changes to existing Resource / Data Source

- [ ] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [ ] I have written new tests for my resource or datasource changes & updated any relevant documentation.
- [ ] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.


## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

<!-- Please include testing logs or evidence here or an explanation on why no testing evidence can be provided. 

For state migrations please test the changes locally and provide details here, such as the versions involved in testing the migration path. For further details on testing state migration changes please see our guide on [state migrations](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-state-migrations.md#testing) in the contributor documentation. -->
```

PASS: TestAccComputeFleet_virtualMachineProfileImage_imageFromSharedImageGalleryVersion (1674.34s)
PASS: TestAccComputeFleet_virtualMachineProfileImage_imageFromSharedImageGallery (1618.08s)
PASS: TestAccComputeFleet_virtualMachineProfileImage_imageFromCommunitySharedImageGalleryVersion (1720.16s)
PASS: TestAccComputeFleet_virtualMachineProfileImage_imageFromCommunitySharedImageGallery (1772.22s)
PASS: TestAccComputeFleet_virtualMachineProfileImage_imageFromImageId (688.07s)
PASS: TestAccComputeFleet_virtualMachineProfileImage_imageFromImageSourceReference (583.79s)
PASS: TestAccComputeFleet_virtualMachineProfileExtensions_complete (238.23s)
PASS: TestAccComputeFleet_virtualMachineProfileExtensions_protectedSettingsFromKeyVault (768.25s)
PASS: TestAccComputeFleet_virtualMachineProfileExtensions_basic (233.85s)
PASS: TestAccComputeFleet_virtualMachineProfileExtensions_update (548.34s)
PASS: TestAccComputeFleet_virtualMachineProfileAuth_authMultipleSSHPublicKeys (616.67s)
PASS: TestAccComputeFleet_virtualMachineProfileAuth_authSSHKey (612.78s)
PASS: TestAccComputeFleet_virtualMachineProfileAuth_authPassword (491.86s)
PASS: TestAccComputeFleet_virtualMachineProfileAuth_authEd25519SSHPublicKeys (446.14s)
PASS: TestAccComputeFleet_virtualMachineProfileAuth_authSSHKeyAndPassword (433.36s)
PASS: TestAccComputeFleet_completeExceptVMSS (323.73s)
PASS: TestAccComputeFleet_update (503.19s)
PASS: TestAccComputeFleet_requiresImport (304.98s)
PASS: TestAccComputeFleet_basic (276.35s)
PASS: TestAccComputeFleet_virtualMachineProfileOthers_UserData (544.01s)
PASS: TestAccComputeFleet_virtualMachineProfileOthers_scheduledEvent (550.24s)
PASS: TestAccComputeFleet_virtualMachineProfileOthers_licenseType (560.89s)
PASS: TestAccComputeFleet_virtualMachineProfileOthers_bootDiagnosticEnabled (537.74s)
PASS: TestAccComputeFleet_virtualMachineProfileOthers_galleryApplication (1636.13s)
PASS: TestAccComputeFleet_virtualMachineProfileOthers_securityProfile (409.62s)
PASS: TestAccComputeFleet_virtualMachineProfileOthers_additionalCapabilitiesHibernation (359.61s)
PASS: TestAccComputeFleet_virtualMachineProfileOthers_additionalCapabilitiesUltraSSD (357.57s)
PASS: TestAccComputeFleet_virtualMachineProfileOsProfile_windowsUpdate (774.72s)
PASS: TestAccComputeFleet_virtualMachineProfileOsProfile_windowsComplete (623.81s)
PASS: TestAccComputeFleet_virtualMachineProfileOsProfile_linuxBasic (322.43s)
PASS: TestAccComputeFleet_virtualMachineProfileOsProfile_linuxUpdate (1631.18s)
PASS: TestAccComputeFleet_virtualMachineProfileOsProfile_windowsBasic (360.23s)
PASS: TestAccComputeFleet_virtualMachineProfileOsProfile_linuxComplete (532.94s)
PASS: TestAccComputeFleet_virtualMachineProfileOsDisk_complete (536.28s)
PASS: TestAccComputeFleet_virtualMachineProfileOsDisk_update (1235.23s)
PASS: TestAccComputeFleet_virtualMachineProfileOsDisk_basic (330.43s)
PASS: TestAccComputeFleet_virtualMachineProfileNetwork_basic (379.29s)
PASS: TestAccComputeFleet_virtualMachineProfileNetwork_updateForBaseVirtualMachineProfile (1878.72s)
PASS: TestAccComputeFleet_virtualMachineProfileNetwork_updateForAdditionalLocation (2297.68s)
PASS: TestAccComputeFleet_virtualMachineProfileNetwork_completeForBaseVirtualMachineProfile (836.75s)
PASS: TestAccComputeFleet_virtualMachineProfileNetwork_completeForAdditionalLocation (848.91s)
PASS: TestAccComputeFleet_virtualMachineProfileDataDisk_basic (360.75s)
PASS: TestAccComputeFleet_virtualMachineProfileDataDisk_complete (469.06s)
PASS: TestAccComputeFleet_virtualMachineProfileDataDisk_update (1296.47s)
```

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

<!-- Replace the changelog example below with your entry. One resource per line. -->

* `azurerm_compute_fleet` - new resource


<!-- What type of PR is this? -->
This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change

sinbai committed:
support new resource compute fleet

ms-zhenhua commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `additional_unattend_content` - (Optional) One or more `additional_unattend_content` blocks as defined above.
+
+* `automatic_updates_enabled` - (Optional) Should the automatic updates of the virtual machines be enabled? Defaults to `true`.
```
```suggestion
* `automatic_updates_enabled` - (Optional) Whether to enable the automatic updates of the virtual machines. Defaults to `true`.
```

sinbai replied:
Fixed.

ms-zhenhua commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `automatic_updates_enabled` - (Optional) Should the automatic updates of the virtual machines be enabled? Defaults to `true`.
+
+* `bypass_platform_safety_checks_enabled` - (Optional) Should the customer to schedule patching without accidental upgrades be enabled? Defaults to `false`.
```
same as `linux_configuration`

sinbai replied:
Fixed.

ms-zhenhua commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `bypass_platform_safety_checks_enabled` - (Optional) Should the customer to schedule patching without accidental upgrades be enabled? Defaults to `false`.
+
+* `hot_patching_enabled` - (Optional) Should the customers to patch the virtual machines without requiring a reboot be enabled? Defaults to `false`.
```
```suggestion
* `hot_patching_enabled` - (Optional) Whether to enable the customers to patch the virtual machines without requiring a reboot. Defaults to `false`.
```

sinbai replied:
Fixed.

ms-zhenhua commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `patch_mode` - (Optional)  Specifies the mode of in-guest patching of the virtual machines. Possible values are `AutomaticByOS`, `AutomaticByPlatform` and `Manual`.
+
+* `provision_vm_agent_enabled` - (Optional) Should the virtual machine agent be provisioned on each virtual machine in the Scale Set? Defaults to `true`.
```
same as `linux_configuration`

sinbai replied:
Fixed.

ms-zhenhua commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `time_zone` - (Optional) Specifies the time zone of the windows virtual machine. Changing this forces a new resource to be created.
+
+* `vm_agent_platform_updates_enabled` - (Optional) Should the virtual machine agent platform updates be enabled for the windows virtual machine? Defaults to `false`.
```
same as `linux_configuration`

sinbai replied:
Fixed.

ms-zhenhua reviewed (changes requested):
Hi @sinbai ,

Thanks for this PR - I've taken a look through and left some comments. If we can fix those up, this should be good to go 👍

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1821 @@
+			ForceNew: true,
+		},
+
+		"vm_attributes": vmAttributesSchema(),
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1821 @@
+						Optional: true,
+						ForceNew: true,
+						Default:  string(fleets.RegularPriorityAllocationStrategyLowestPrice),
+						ValidateFunc: validation.StringInSlice([]string{
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1821 @@
+						Optional: true,
+						ForceNew: true,
+						Default:  string(fleets.SpotAllocationStrategyPriceCapacityOptimized),
+						ValidateFunc: validation.StringInSlice([]string{
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1821 @@
+						Optional: true,
+						ForceNew: true,
+						Default:  string(fleets.EvictionPolicyDelete),
+						ValidateFunc: validation.StringInSlice([]string{
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1821 @@
+					}
+				}
+
+				if a := model.AdditionalLocationProfile; len(a) > 0 && len(a[0].VirtualMachineProfileOverride) > 0 {
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1821 @@
+			// API requires `osProfile.adminPassword` when updating resource but the GET API does not return the sensitive data `osProfile.adminPassword`
+			if props := properties.Properties; props != nil {
+				if len(model.VirtualMachineProfile[0].OsProfile[0].LinuxConfiguration) > 0 {
+					if v := props.ComputeProfile.BaseVirtualMachineProfile.OsProfile; v != nil {
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1821 @@
+
+				if a := model.AdditionalLocationProfile; len(a) > 0 && len(a[0].VirtualMachineProfileOverride) > 0 {
+					if os := a[0].VirtualMachineProfileOverride[0].OsProfile; len(os) > 0 && len(os[0].WindowsConfiguration) > 0 {
+						if v := props.AdditionalLocationsProfile; v != nil && len(v.LocationProfiles) > 0 {
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1821 @@
+			}
+
+			if v := state.AdditionalLocationProfile; len(v) > 0 && len(v[0].VirtualMachineProfileOverride) > 0 {
+				osProfile = state.AdditionalLocationProfile[0].VirtualMachineProfileOverride[0].OsProfile[0]
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1821 @@
+				return err
+			}
+
+			if len(state.AdditionalLocationProfile) > 0 {
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1821 @@
+				}
+			}
+			for _, v := range state.VirtualMachineProfile[0].Extension {
+				if v.ProtectedSettingsJson != "" {
+					if len(v.ProtectedSettingsFromKeyVault) > 0 {
+						return fmt.Errorf("`protected_settings_from_key_vault` cannot be used with `protected_settings_json`")
+					}
+				}
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1821 @@
+	if len(input.ExcludedVMSizes) > 0 {
+		output.ExcludedVMSizes = pointer.To(input.ExcludedVMSizes)
+	}
+	output.MemoryInGiB = pointer.From(expandVMAttributeMinMaxDoubleModel(input.MemoryInGib))
+
+	output.VCPUCount = pointer.From(expandVMAttributeMinMaxIntegerModel(input.VCPUCount))
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1821 @@
+		return outputList
+	}
+	for _, input := range inputList {
+		output := pointer.From(input.Id)
+		outputList = append(outputList, output)
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1821 @@
+	}
+
+	for _, input := range *inputList {
+		output := IPTagModel{}
+
+		output.Type = pointer.From(input.IPTagType)
+		output.Tag = pointer.From(input.Tag)
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1821 @@
+	}
+	output.MaxHourlyPricePerVM = maxHourlyPricePerVM
+
+	output.MinCapacity = pointer.From(input.MinCapacity)
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1821 @@
+		VMCategories:              flattenToStringSlice(input.VMCategories),
+	}
+
+	output.AcceleratorSupport = string(pointer.From(input.AcceleratorSupport))
+	output.BurstableSupport = string(pointer.From(input.BurstableSupport))
+	output.ExcludedVMSizes = pointer.From(input.ExcludedVMSizes)
+	output.LocalStorageSupport = string(pointer.From(input.LocalStorageSupport))
+	output.RdmaSupport = string(pointer.From(input.RdmaSupport))
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1821 @@
+	if input == nil {
+		return outputList
+	}
+	output := VMAttributeMinMaxIntegerModel{}
+	output.Max = pointer.From(input.Max)
+	output.Min = pointer.From(input.Min)
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1821 @@
+	if input == nil {
+		return outputList
+	}
+	output := VMAttributeMinMaxDoubleModel{}
+	output.Max = pointer.From(input.Max)
+	output.Min = pointer.From(input.Min)
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1821 @@
+	}
+
+	for _, input := range *inputList {
+		output := VMSizeProfileModel{
+			Name: input.Name,
+		}
+		output.Rank = pointer.From(input.Rank)
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2650 @@
+		MaxItems: 1,
+		Elem: &pluginsdk.Resource{
+			Schema: map[string]*pluginsdk.Schema{
+				"os_profile": osProfileSchema(),
+
+				"network_api_version": {
+					Type:         pluginsdk.TypeString,
+					Required:     true,
+					ValidateFunc: validation.StringIsNotEmpty,
+				},
+
+				"network_interface": networkInterfaceSchema(),
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2650 @@
+					ValidateFunc: validation.StringIsNotEmpty,
+				},
+
+				"capacity_reservation_group_id": {
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2650 @@
+					Default:  false,
+				},
+
+				"network_security_group_id": {
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2650 @@
+					ValidateFunc: networksecuritygroups.ValidateNetworkSecurityGroupID,
+				},
+
+				"primary": {
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2650 @@
+							"admin_username": {
+								Type:         pluginsdk.TypeString,
+								Required:     true,
+								ValidateFunc: validateAdminUsernameLinux,
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2650 @@
+								Type:         pluginsdk.TypeString,
+								Optional:     true,
+								Sensitive:    true,
+								ValidateFunc: validatePasswordComplexityLinux,
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2650 @@
+							"admin_username": {
+								Type:         pluginsdk.TypeString,
+								Required:     true,
+								ValidateFunc: validateAdminUsernameWindows,
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2650 @@
+				"create_option": {
+					Type:     pluginsdk.TypeString,
+					Required: true,
+					ValidateFunc: validation.StringInSlice([]string{
```
Because the possible values ​​include the following const, but only `Empty` and `FromImage` are supported here. So it is not possible.
![image](https://github.com/user-attachments/assets/0d1eee85-fa9c-4416-84b1-1e306c394b0b)

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2650 @@
+	}
+
+	if isAdditional {
+		encryptionAtHostEnabledExist := d.GetRawConfig().AsValueMap()["additional_location_profile"].AsValueSlice()[0].AsValueMap()["virtual_machine_profile_override"].AsValueSlice()[0].AsValueMap()["encryption_at_host_enabled"]
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2650 @@
+	}
+
+	if isAdditional {
+		encryptionAtHostEnabledExist := d.GetRawConfig().AsValueMap()["additional_location_profile"].AsValueSlice()[0].AsValueMap()["virtual_machine_profile_override"].AsValueSlice()[0].AsValueMap()["encryption_at_host_enabled"]
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2650 @@
+		rebootSetting := v[0].RebootSetting
+		bypassPlatformSafetyChecksEnabledExist := d.GetRawConfig().AsValueMap()["virtual_machine_profile"].AsValueSlice()[0].AsValueMap()["os_profile"].AsValueSlice()[0].AsValueMap()["windows_configuration"].AsValueSlice()[0].AsValueMap()["bypass_platform_safety_checks_enabled"]
+		if isAdditional {
+			bypassPlatformSafetyChecksEnabledExist = d.GetRawConfig().AsValueMap()["additional_location_profile"].AsValueSlice()[0].AsValueMap()["virtual_machine_profile_override"].AsValueSlice()[0].AsValueMap()["os_profile"].AsValueSlice()[0].AsValueMap()["windows_configuration"].AsValueSlice()[0].AsValueMap()["bypass_platform_safety_checks_enabled"]
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2650 @@
+		rebootSetting := v[0].RebootSetting
+		bypassPlatformSafetyChecksEnabledExist := d.GetRawConfig().AsValueMap()["virtual_machine_profile"].AsValueSlice()[0].AsValueMap()["os_profile"].AsValueSlice()[0].AsValueMap()["linux_configuration"].AsValueSlice()[0].AsValueMap()["bypass_platform_safety_checks_enabled"]
+		if isAdditional {
+			bypassPlatformSafetyChecksEnabledExist = d.GetRawConfig().AsValueMap()["additional_location_profile"].AsValueSlice()[0].AsValueMap()["virtual_machine_profile_override"].AsValueSlice()[0].AsValueMap()["os_profile"].AsValueSlice()[0].AsValueMap()["linux_configuration"].AsValueSlice()[0].AsValueMap()["bypass_platform_safety_checks_enabled"]
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2650 @@
+
+	for _, input := range *input.GalleryApplications {
+		output := GalleryApplicationModel{}
+		output.VersionId = input.PackageReferenceId
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2650 @@
+		output.CustomDataBase64 = pointer.From(input.CustomData)
+	} else {
+		if isAdditional {
+			output.CustomDataBase64 = utils.Base64EncodeIfNot(d.Get("additional_location_profile.0.virtual_machine_profile_override.0.os_profile.0.custom_data_base64").(string))
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2650 @@
+			Secret:                        flattenOsProfileWindowsSecretsModel(input.Secrets),
+		}
+		if isAdditional {
+			windowsConfig.AdminPassword = d.Get("additional_location_profile.0.virtual_machine_profile_override.0.os_profile.0.windows_configuration.0.admin_password").(string)
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2650 @@
+		}
+
+		if isAdditional {
+			linuxConfig.AdminPassword = d.Get("additional_location_profile.0.virtual_machine_profile_override.0.os_profile.0.linux_configuration.0.admin_password").(string)
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2650 @@
+		output := AdditionalUnattendContentModel{}
+		existing := make([]interface{}, 0)
+		if isAdditional {
+			if v, ok := d.GetOk("additional_location_profile.0.virtual_machine_profile_override.0.os_profile.0.windows_configuration.0.additional_unattend_content"); ok {
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2650 @@
+			output.ForceExtensionExecutionOnChange = pointer.From(props.ForceUpdateTag)
+			// Sensitive data isn't returned, so we get it from config
+			if isAdditional {
+				output.ProtectedSettingsJson = metadata.ResourceData.Get("additional_location_profile.0.virtual_machine_profile_override.0.extension." + strconv.Itoa(i) + ".protected_settings_json").(string)
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `virtual_machine_profile` - (Required) A `virtual_machine_profile` block as defined below. Changing this forces a new resource to be created.
+
+* `additional_capabilities_hibernation_enabled` - (Optional) Should the hibernation capability to be enabled on the Compute Fleet?  Defaults to `false`. Changing this forces a new resource to be created.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `additional_capabilities_hibernation_enabled` - (Optional) Should the hibernation capability to be enabled on the Compute Fleet?  Defaults to `false`. Changing this forces a new resource to be created.
+
+* `additional_capabilities_ultra_ssd_enabled` - (Optional) Should the capacity to enable Data Disks of the `UltraSSD_LRS` storage account type be supported on this Compute Fleet? Defaults to `false`. Changing this forces a new resource to be created.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `vm_attributes` - (Optional) A `vm_attributes` block as defined below. Changing this forces a new resource to be created.
+
+* `zones` - (Optional) Specifies a list of availability zones which the Compute Fleet is available. Changing this forces a new resource to be created.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `tags` - (Optional) A mapping of tags which should be assigned to the Compute Fleet.
+
+* `vm_sizes_profile` - (Optional) One or more `vm_sizes_profile` blocks as defined below. Conflicts with `vm_attributes.0.excluded_vm_sizes`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `network_interface` - (Required) One or more `network_interface` blocks as defined above.
+
+* `boot_diagnostic_enabled` - (Optional) Should the boot diagnostics be enabled on the virtual machine? Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `data_disk` - (Optional) One or more `data_disk` blocks as defined above.
+
+* `encryption_at_host_enabled` - (Optional) Should disks attached to the Compute Fleet be encrypted by enabling encryption at host? Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `extension` - (Optional) One or more `extension` blocks as defined above.
+
+* `extension_operations_enabled` - (Optional) Should extension operations be allowed on the Compute Fleet?  Defaults to `true`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `scheduled_event_termination_timeout` - (Optional) Specifies the length of time a virtual machine being re-imaged or having its OS upgraded will have to potentially approve the OS image scheduled event before the event is auto approved (timed out). The configuration is specified in ISO 8601 format. Possible values are `PT5M` and `PT15M`.
+
+* `secure_boot_enabled` - (Optional) Should the secure boot be enabled the virtual machine? Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `user_data_base64` - (Optional) The base64-encoded User Data which should be used for the Compute Fleet.
+
+* `vtpm_enabled` - (Optional) Should the vTPM be enabled on the virtual machine? Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `ip_configuration` - (Required) One or more `ip_configuration` blocks as defined above.
+
+* `accelerated_networking_enabled` - (Optional) Should the network interface support accelerated networking be enabled? Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `dns_servers` - (Optional) Specifies a list of IP addresses of DNS servers which should be assigned to the network interface.
+
+* `ip_forwarding_enabled` - (Optional) Should the network interface support IP forwarding be enabled? Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `disk_encryption_set_id` - (Optional) The ID of the disk encryption set which should be used to encrypt the data disk.
+
+* `disk_size_in_gb` - (Optional) The size of the data disk which should be created. Required if `create_option` is specified as `Empty`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `disk_size_in_gb` - (Optional) The size of the data disk which should be created. Required if `create_option` is specified as `Empty`.
+
+* `lun` - (Optional) The logical unit number of the data disk, which must be unique within the virtual machine. Required if `create_option` is specified as `Empty`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `storage_account_type` - (Optional)  The type of storage account which should back the data disk. Possible values include `Premium_LRS`, `PremiumV2_LRS`, `Premium_ZRS`, `Standard_LRS`, `StandardSSD_LRS`, `StandardSSD_ZRS` and `UltraSSD_LRS`.
+
+* `write_accelerator_enabled` - (Optional) Should the write accelerator be enabled on the data disk? Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `type_handler_version` - (Required) Specifies the version of the extension to use, available versions can be found using the Azure CLI.
+
+* `auto_upgrade_minor_version_enabled` - (Optional) Should the latest version of the extension be used at deployment time, if one is available? This won't auto-update the extension on existing installation. Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `auto_upgrade_minor_version_enabled` - (Optional) Should the latest version of the extension be used at deployment time, if one is available? This won't auto-update the extension on existing installation. Defaults to `false`.
+
+* `automatic_upgrade_enabled` - (Optional) Should the extension be automatically upgraded by the platform, if there is a newer version of the extension available? Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `extensions_to_provision_after_vm_creation` - (Optional) Specifies an ordered list of extension names which Compute Fleet should provision after virtual machine creation.
+
+* `failure_suppression_enabled` - (Optional) Should the failures from the extension be suppressed? Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `version_id` - (Required) Specifies the gallery application version resource ID.
+
+* `automatic_upgrade_enabled` - (Optional) Should the gallery application be automatically upgraded, if there is a newer version of the gallery application available? Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `tag` - (Optional) Specifies a passthrough value for more generic context.
+
+* `treat_failure_as_deployment_failure_enabled` - (Optional) Should any failure for any operation in the virtual machine application will fail the deployment be enabled? Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `load_balancer_backend_address_pool_ids` - (Optional) Specifies a list of backend address pools IDs from a load balancer which the Compute Fleet should be connected to.
+
+* `primary` - (Optional) Is this the primary IP configuration? Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `admin_ssh_keys` - (Optional) Specifies a list of the public key which should be used for authentication, which needs to be in `ssh-rsa` format with at least 2048-bit or in `ssh-ed25519` format.
+
+* `bypass_platform_safety_checks_enabled` - (Optional) Should the customer to schedule patching without accidental upgrades be enabled? Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `bypass_platform_safety_checks_enabled` - (Optional) Should the customer to schedule patching without accidental upgrades be enabled? Defaults to `false`.
+
+* `password_authentication_enabled` - (Optional) When an `admin_password` is specified `password_authentication_enabled` must be set to `true`. Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `patch_mode` - (Optional)  Specifies the mode of in-guest patching of the virtual machines. Possible values are `AutomaticByPlatform` and `ImageDefault`.
+
+* `provision_vm_agent_enabled` - (Optional) Should the virtual machine agent be provisioned on each virtual machine in the Scale Set? Defaults to `true`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `secret` - (Optional) One or more `secret` blocks as defined below.
+
+* `vm_agent_platform_updates_enabled` - (Optional) Should the virtual machine agent be provisioned on each linux virtual machine in the Compute Fleet? Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `ip_configuration` - (Required) One or more `ip_configuration` blocks as defined above.
+
+* `accelerated_networking_enabled` - (Optional) Should the network interface support accelerated networking be enabled? Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `dns_servers` - (Optional) Specifies a list of IP addresses of DNS servers which should be assigned to the network interface.
+
+* `ip_forwarding_enabled` - (Optional) Should the network interface support IP forwarding be enabled? Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `network_security_group_id` - (Optional) The ID of the network security group which should be assigned to the network interface.
+
+* `primary` - (Optional) Is this the primary network interface? Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `storage_account_type` - (Optional) The Type of Storage Account which should back the OS Disk. Possible values include `Premium_LRS`, `Premium_ZRS`, `Standard_LRS`, `StandardSSD_LRS` and `StandardSSD_ZRS`.
+
+* `write_accelerator_enabled` - (Optional) Should the write accelerator be enabled on the OS Disk? Defaults to `false`.
```
Fixed.

sinbai commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,814 @@
+
+* `eviction_policy` - (Optional) The policy which should be used by spot virtual machines that are evicted from the Compute Fleet. Defaults to `Delete`. Possible values are `Deallocate` and `Delete`. Changing this forces a new resource to be created.
+
+* `maintain_enabled` - (Optional) Should the continuous goal seeking for the desired capacity and restoration of evicted spot virtual machines be enabled? Defaults to `true`. Changing this forces a new resource to be created.
```
Fixed.

sinbai committed:
fix comments

sinbai commented:
> Hi @sinbai ,
> 
> Thanks for this PR - I've taken a look through and left some comments. If we can fix those up, this should be good to go 👍

Hi @ms-zhenhua thank you very much for your feedback. I have updated the code, could you please take another look?

ms-zhenhua commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -890,29 +876,31 @@ func (r ComputeFleetResource) Update() sdk.ResourceFunc {
 			if props := properties.Properties; props != nil {
-				if len(model.VirtualMachineProfile[0].OsProfile[0].LinuxConfiguration) > 0 {
-					if v := props.ComputeProfile.BaseVirtualMachineProfile.OsProfile; v != nil {
+				if v := props.ComputeProfile.BaseVirtualMachineProfile.OsProfile; v != nil {
```
what if `ComputeProfile == nil` or `BaseVirtualMachineProfile == nil` ?

sinbai replied:
`ComputeProfile` and `BaseVirtualMachineProfile` are not pointer types, so I assume they will not be equal to nil.

ms-zhenhua reviewed (changes requested):
Hi @sinbai,

Thank you for the updates. Only one comment left. Others LGTM~

ms-zhenhua reviewed (dismissed):
LGTM~

diyellap commented:
@sinbai Can we remove all the examples that update ComputeProfile on Fleet update in this PR? Updating ComputeProfile is being blocked by Fleet. Fleet creation with ComputeProfile is still valid.

sinbai commented:
Due to API changes, this PR was converted to a draft PR until the code update is complete.

sinbai committed:
add forcenew for all properties of virtual machine config.

katbyte commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,740 @@
+
+* `os_disk` - (Optional) A `os_disk` block as defined above. Changing this forces a new resource to be created.
+
+* `scheduled_event_os_image_timeout` - (Optional) Specifies the length of time a virtual machine being deleted will have to potentially approve the terminate scheduled event before the event is auto approved (timed out). The configuration must be specified in ISO 8601 format. The only possible value is `PT15M`. Changing this forces a new resource to be created.
+
+* `scheduled_event_termination_timeout` - (Optional) Specifies the length of time a virtual machine being reimaged or having its OS upgraded will have to potentially approve the OS image scheduled event before the event is auto approved (timed out). The configuration is specified in ISO 8601 format. Possible values are `PT5M` and `PT15M`. Changing this forces a new resource to be created.
```
these are ISO durations? so should they be
```suggestion
* `scheduled_event_os_image_timeout_duration` - (Optional) Specifies the length of time a virtual machine being deleted will have to potentially approve the terminate scheduled event before the event is auto approved (timed out). The configuration must be specified in ISO 8601 format. The only possible value is `PT15M`. Changing this forces a new resource to be created.

* `scheduled_event_termination_timeout_duration` - (Optional) Specifies the length of time a virtual machine being reimaged or having its OS upgraded will have to potentially approve the OS image scheduled event before the event is auto approved (timed out). The configuration is specified in ISO 8601 format. Possible values are `PT5M` and `PT15M`. Changing this forces a new resource to be created.
```
to make it clear its not seconds etc

katbyte commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,740 @@
+
+---
+
+A `additional_unattend_content` block supports the following:
+
+* `content` - (Required) Specifies the XML formatted content that is added to the unattend.xml file for the specified path and component. Changing this forces a new resource to be created.
```
content seems redundant here?

```suggestion
A `additional_unattend_content` block supports the following:

* `xml` - (Required) Specifies the XML formatted content that is added to the unattend.xml file for the specified path and component. Changing this forces a new resource to be created.
```
might make more sense?

sinbai replied:
Fixed.

katbyte commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,740 @@
+
+---
+
+A `ip_tag` block supports the following:
+
+* `tag` - (Required) The IP tag associated with the public IP. Changing this forces a new resource to be created.
+
+* `type` - (Required) The type of IP tag. Changing this forces a new resource to be created.
```
type? what are the possible values of type?

i would have expected it to be tag.key and tag.value?

sinbai replied:
I think it makes sense to understand that "ipTagType is the key of the tag, and tag is the value", but from the perspective of Azure's data structure semantics and API design intent, they are not key-value pairs in the traditional sense, but a combination of `purpose category` + `purpose name`.

For example:
```
"ipTags": [
{
"ipTagType": "RoutingPreference",
"tag": "Internet"
}
]

"ipTags": [
{
"ipTagType": "RoutingPreference",
"tag": "Microsoft"
}
]
```
The above IP addresses uses routing preference (RoutingPreference), specifically Internet/Microsoft routing.

Therefore, I assume that these two sets of fields together describe the multiple dimensional uses of this IP, rather than a pure tags mapping. What about you?

katbyte replied:
Well is any key/value possible or is there only a subset that are valid?

sinbai replied:
Hi @SahilArora92 for the question above, "Well is any key/value possible or is there only a subset that are valid?", could you please help me check it?

If it is not any key/value possible, could you please help me confirm all possible values ​​​​for [ipTagType](https://github.com/Azure/azure-rest-api-specs/blob/0bf123645e53f9006c15a962087c9073d7fd455f/specification/azurefleet/resource-manager/Microsoft.AzureFleet/stable/2024-11-01/azurefleet.json#L3009) and [tag](https://github.com/Azure/azure-rest-api-specs/blob/0bf123645e53f9006c15a962087c9073d7fd455f/specification/azurefleet/resource-manager/Microsoft.AzureFleet/stable/2024-11-01/azurefleet.json#L3013) so that I could write it into the document? Thanks in advance!

ykh015 replied:
Hi @sinbai There is no specific key/value combinations here. These is a dynamic key-value pair.

IPTagType can have values like NetworkDomain, FirstPartyUsage, RoutingPreference etc.

sinbai replied:
> Hi @sinbai There is no specific key/value combinations here. These is a dynamic key-value pair.
> 
> IPTagType can have values like NetworkDomain, FirstPartyUsage, RoutingPreference etc.

Hi @ykh015 thanks for your response, can I confirm that you are a member of the service team, right? Can I consider this response official?

ykh015 replied:
Yes, I am the member of the service team.

sinbai replied:
Hi @katbyte could you please take a look at the response from the service team? If there are any changes needed, please feel free to let me know. Thank you!

katbyte commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,740 @@
+
+* `provision_vm_agent_enabled` - (Optional) Whether to provision the virtual machine agent on each virtual machine in the Scale Set. Defaults to `true`. Changing this forces a new resource to be created.
+
+* `reboot_setting` - (Optional) Specifies the reboot setting for all `AutomaticByPlatform` patch installation operations. Possible values are `Always`, `IfRequired`, `Never` and `Unknown`. Changing this forces a new resource to be created.
```
as this si for patching
```suggestion
* `patch_rebooting` - (Optional) Specifies the reboot setting for all `AutomaticByPlatform` patch installation operations. Possible values are `Always`, `IfRequired`, `Never` and `Unknown`. Changing this forces a new resource to be created.
```

katbyte commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,740 @@
+
+* `disk_encryption_set_id` - (Optional) The ID of the Disk Encryption Set which should be used to encrypt the OS Disk. Changing this forces a new resource to be created.
+
+* `disk_size_in_gb` - (Optional) The size of the internal OS Disk in GB, if you wish to vary from the size used in the image the Compute Fleet is sourced from. Changing this forces a new resource to be created.
```
is this gb or gib?

memory_in_gib_per_vcpu is used elsewhere in this resource so is this different and not gib?

sinbai replied:
After confirmation, the unit of disk/os size should be GiB. Fixed.

katbyte commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,740 @@
+
+-> **Note:** Once the `burstable_support` has been specified, removing it forces a new resource to be created.
+
+* `cpu_manufacturers` - (Optional) Specifies a list of the virtual machine CPU manufacturers. Possible values are `AMD`, `Ampere`, `Intel` and `Microsoft`.
+
+-> **Note:** Once the `cpu_manufacturers` has been specified, removing it forces a new resource to be created.
```
i think we can combine these:
```suggestion
* `cpu_manufacturers` - (Optional) Specifies a list of the virtual machine CPU manufacturers. Possible values are `AMD`, `Ampere`, `Intel` and `Microsoft`. Once specified changing this value forces a new resource to be created.
```
?

sinbai replied:
Fixed.

katbyte commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1761 @@
+			properties.Properties.AdditionalLocationsProfile = additionalLocationsProfileValue
+
+			computeProfile := fleets.ComputeProfile{
+				AdditionalVirtualMachineCapabilities: &fleets.AdditionalCapabilities{
+					HibernationEnabled: pointer.To(model.AdditionalCapabilitiesHibernationEnabled),
+					UltraSSDEnabled:    pointer.To(model.AdditionalCapabilitiesUltraSSDEnabled),
+				},
```
could we create a block in schema to match the API as well improve ux?

sinbai replied:
Yes, fixed.

katbyte commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,740 @@
+
+* `extension_operations_enabled` - (Optional) Whether to enable extension operations on the Compute Fleet.  Defaults to `true`. Changing this forces a new resource to be created.
+
+* `extensions_time_budget` - (Optional) Specifies the time alloted for all extensions to start. Changing this forces a new resource to be created.
```
what unit is this in? seconds? min? duration?

sinbai replied:
The unit should be duration, fixed.

sinbai committed:
fix comments

sinbai committed:
update code

mbfrahry reviewed (commented):
It looks like this service is still in Preview. Do we have an idea of when it'll be in GA. We're hesitant to merge in Services that are in Preview as they've changed out from underneath us when moving from Preview to GA which causes a poor user experience and makes it difficult to maintain
<img width="706" alt="image" src="https://github.com/user-attachments/assets/33687f0e-52f7-474d-86e9-07a758b30aea" />

ykh015 commented:
Hi @mbfrahry Azure Compute Fleet is GA as of April end.

Here is the announcement:
https://techcommunity.microsoft.com/blog/azurecompute/azure-compute-fleet---generally-available/4408859

The documentation is dated and will be updated but the product has GA support

mbfrahry reviewed (changes requested):
This is a massive PR and it'll have to be reviewed in pieces. The first part is just focused on the schema. 

It looks like there are two major parts of this that are in preview that we shouldn't support in case they change out from underneath us. 

The other bits are just general pieces to make the user experience better.

sinbai committed:
Remove preview features

sinbai committed:
Remove preview features

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1802 @@
+
+		"zones": commonschema.ZonesMultipleOptionalForceNew(),
+
+		"compute_api_version": {
```
It looks like the `compute_api_version` property is not marked as `Computed`. Could you kindly take another look to confirm?

mbfrahry replied:
I'm asking if it should be marked as Computed since you're using `ignore_changes` in the tests https://github.com/hashicorp/terraform-provider-azurerm/pull/28758/files#diff-1f8c8b4f4081dc977d64b3c296f68407efe9b67bd219376e7869fc5bc79d3bd3R292

sinbai replied:
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1802 @@
+			ForceNew: true,
+		},
+
+		"vm_attributes": {
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1802 @@
+
+		"additional_capabilities": virtualMachineAdditionalCapabilitiesSchema(),
+
+		"additional_location_profile": {
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1802 @@
+						ValidateFunc: validation.StringInSlice(fleets.PossibleValuesForEvictionPolicy(), false),
+					},
+
+					"maintain_enabled": {
```
Since `maintain_capacity` is a boolean, would it make sense to rename it to `maintain_capacity_enabled`?

mbfrahry replied:
Sure, that'll work

wyattfry replied:
I see the update in the documentation, but still have stragglers here, could you update them all to `maintain_capacity_enabled`?

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1802 @@
+
+		"vm_sizes_profile": {
+			Type:     pluginsdk.TypeList,
+			Optional: true,
```
Fixed.

sinbai commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1802 @@
+					"min_capacity": {
+						Type:     pluginsdk.TypeInt,
+						ForceNew: true,
+						Optional: true,
```
As both the API and the Portal allow `min_capacity` to be 0, I added `ValidateFunc: validation.IntAtLeast(0)` and set default to `0`.

The attribute remains optional. Does this approach look reasonable to you?

mbfrahry replied:
That'll work!

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2654 @@
+					ValidateFunc: validation.StringIsBase64,
+				},
+
+				"linux_configuration": {
```
No, it's not allowed to specify both `linux_configuration` and `windows_configuration` in the same `os_profile`. As the preview feature `additional_location_profile` has been removed, I replaeced the `CustomizeDiff` check (`only one of `linux_configuration` and `windows_configuration` in `virtual_machine_profile` must be specified`) with the `ExactlyOneOf` schema constraint instead.

mbfrahry replied:
Perfect!

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2654 @@
+		MaxItems: 100,
+		Elem: &pluginsdk.Resource{
+			Schema: map[string]*pluginsdk.Schema{
+				"version_id": {
```
I have tested this in the `TestAccComputeFleet_virtualMachineProfileOthers_galleryApplication` test case.
When I updated the test to specify only the `version_id` of the `gallery_application`, it passed successfully.

sinbai committed:
fix comments

sinbai committed:
fix acctests failure

remove zone 2

sinbai committed:
remove redundant resources related to additional_location_profile

sinbai committed:
fix acctest failure and comments

mbfrahry commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1802 @@
+						Type:         pluginsdk.TypeFloat,
+						Optional:     true,
+						ForceNew:     true,
+						Default:      -1,
```
Ah I see. I was looking at the way it was before your most recent changes and `-1` would not have been picked up. 

```
if input.MaxHourlyPricePerVM > 0 {
	output.MaxPricePerVM = pointer.To(input.MaxHourlyPricePerVM)
}
```

The new change looks fine except someone could still pass in 0 and I'm not sure what the api/Terraform would do in that instance

mbfrahry commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1802 @@
+
+					"capacity": {
+						Type:         pluginsdk.TypeInt,
+						Optional:     true,
```
That'll work!

sinbai commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2654 @@
+		ForceNew: true,
+		MaxItems: 1,
+		Elem: &pluginsdk.Resource{
+			Schema: map[string]*pluginsdk.Schema{
```
Fixed.

sinbai commented:
> This is a massive PR and it'll have to be reviewed in pieces. The first part is just focused on the schema.
> 
> It looks like there are two major parts of this that are in preview that we shouldn't support in case they change out from underneath us.
> 
> The other bits are just general pieces to make the user experience better.

Hi @mbfrahry thank you very much for your feedback. I’ve addressed all the comments and made the necessary changes. All tests have also passed in TC. Could you please take another look when you get a chance?

sinbai committed:
fix website-lint issue

wyattfry commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1006 @@
+	}
+}
+
+func virtualMachineAdditionalCapabilitiesSchema() *pluginsdk.Schema {
```
Could this schema be moved into the rest of the explicit definition? It seems this function is only called once.

wyattfry commented on internal/services/computefleet/compute_fleet_virtual_machine_profile.go:
```diff
@@ -0,0 +1,2521 @@
+	"github.com/hashicorp/terraform-provider-azurerm/utils"
+)
+
+func virtualMachineProfileSchema() *pluginsdk.Schema {
```
Could this schema be moved into the rest of the explicit definition? It seems this function is only called once.

wyattfry commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,573 @@
+
+* `provision_vm_agent_enabled` - (Optional) Whether to provision the virtual machine agent on each virtual machine in the Scale Set. Defaults to `true`. Changing this forces a new resource to be created.
+
+* `reboot_setting` - (Optional) Specifies the reboot setting for all `AutomaticByPlatform` patch installation operations. Possible values are `Always`, `IfRequired`, `Never` and `Unknown`. Changing this forces a new resource to be created.
```
```suggestion
* `patch_rebooting` - (Optional) Specifies the reboot setting for all `AutomaticByPlatform` patch installation operations. Possible values are `Always`, `IfRequired`, `Never` and `Unknown`. Changing this forces a new resource to be created.
```

wyattfry commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,573 @@
+* `spot_priority_profile` - (Optional) A `spot_priority_profile` block as defined below.
+
+* `tags` - (Optional) A mapping of tags which should be assigned to the Compute Fleet.
+* 
```
```suggestion
```

wyattfry commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,573 @@
+
+A `network_interface` block supports the following:
+
+* `name` - (Required)  The name which should be used for the network interface. Changing this forces a new resource to be created.
```
```suggestion
* `name` - (Required) The name which should be used for the network interface. Changing this forces a new resource to be created.
```

wyattfry commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,573 @@
+
+A `additional_capabilities` block supports the following:
+
+* `hibernation_enabled` - (Optional) Whether to enable the hibernation capability on the Compute Fleet.  Defaults to `false`. Changing this forces a new resource to be created.
```
```suggestion
* `hibernation_enabled` - (Optional) Whether to enable the hibernation capability on the Compute Fleet. Defaults to `false`. Changing this forces a new resource to be created.
```

wyattfry commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,573 @@
+
+* `disk_encryption_set_id` - (Optional) The ID of the disk encryption set which should be used to encrypt the data disk. Changing this forces a new resource to be created.
+
+* `disk_size_in_gib` - (Optional) The size of the data disk which should be created.  Changing this forces a new resource to be created.
```
Maybe indicate the unit once more in the description to emphasize the differentiation from GB?
```suggestion
* `disk_size_in_gib` - (Optional) The size of the data disk which should be created in GiB. Changing this forces a new resource to be created.
```

wyattfry commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,573 @@
+
+-> **Note:** Required if `create_option` is specified as `Empty`.
+
+* `storage_account_type` - (Optional)  The type of storage account which should back the data disk. Possible values include `Premium_LRS`, `PremiumV2_LRS`, `Premium_ZRS`, `Standard_LRS`, `StandardSSD_LRS`, `StandardSSD_ZRS` and `UltraSSD_LRS`. Changing this forces a new resource to be created.
```
```suggestion
* `storage_account_type` - (Optional) The type of storage account which should back the data disk. Possible values include `Premium_LRS`, `PremiumV2_LRS`, `Premium_ZRS`, `Standard_LRS`, `StandardSSD_LRS`, `StandardSSD_ZRS` and `UltraSSD_LRS`. Changing this forces a new resource to be created.
```

wyattfry commented on website/docs/r/compute_fleet.html.markdown:
```diff
@@ -0,0 +1,573 @@
+
+* `vm_agent_platform_updates_enabled` - (Optional) Whether to enable the virtual machine agent platform updates for the windows virtual machine in the Compute Fleet. Defaults to `false`. Changing this forces a new resource to be created.
+
+* `winrm_listener` - (Optional) One or more `winrm_listener` blocks as defined below.  Changing this forces a new resource to be created.
```
```suggestion
* `winrm_listener` - (Optional) One or more `winrm_listener` blocks as defined below. Changing this forces a new resource to be created.
```

wyattfry commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1006 @@
+
+			if len(state.SpotPriorityProfile) > 0 && len(state.RegularPriorityProfile) > 0 {
+				if state.SpotPriorityProfile[0].Capacity+state.RegularPriorityProfile[0].Capacity > 10000 {
+					return fmt.Errorf("the sum of `spot_priority_profile.0.capacity` and `regular_priority_profile.0.capacity` must be between `0` and `10000`, inclusive")
```
When not using a format string for an error message, you can simplify and use `errors.New()` (make sure to import `errors`)
```suggestion
					return errors.New("the sum of `spot_priority_profile.0.capacity` and `regular_priority_profile.0.capacity` must be between `0` and `10000`, inclusive")
```

wyattfry commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1006 @@
+					}
+
+					if len(state.VMSizesProfile) < 3 {
+						return fmt.Errorf("`vm_sizes_profile` must be at least 3 Vm sizes if `spot_priority_profile.0.maintain_enabled` is enabled")
```
```suggestion
						return errors.New("`vm_sizes_profile` must be at least 3 Vm sizes if `spot_priority_profile.0.maintain_enabled` is enabled")
```

wyattfry commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1006 @@
+					}
+
+					if len(state.Zones) == 0 {
+						return fmt.Errorf("enabling `spot_priority_profile.0.maintain_enabled` requires all qualified availability zones in the region to be supported")
```
```suggestion
						return errors.New("enabling `spot_priority_profile.0.maintain_enabled` requires all qualified availability zones in the region to be supported")
```

wyattfry commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1006 @@
+				}
+
+				if state.SpotPriorityProfile[0].MinCapacity > state.SpotPriorityProfile[0].Capacity {
+					return fmt.Errorf("`spot_priority_profile.0.min_capacity` must be between `0` and `spot_priority_profile.0.capacity`, inclusive")
```
```suggestion
					return errors.New("`spot_priority_profile.0.min_capacity` must be between `0` and `spot_priority_profile.0.capacity`, inclusive")
```

wyattfry commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1006 @@
+				}
+
+				if !ultraSSDEnabled && storageAccountType == string(fleets.StorageAccountTypesUltraSSDLRS) {
+					return fmt.Errorf("`UltraSSD_LRS` storage account type can be used only when `ultra_ssd_enabled` is enalbed")
```
```suggestion
					return errors.New("`UltraSSD_LRS` storage account type can be used only when `ultra_ssd_enabled` is enabled")
```

wyattfry commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1006 @@
+
+				if v[0].CreateOption == string(fleets.DiskCreateOptionTypesEmpty) {
+					if v[0].DiskSizeInGiB == 0 {
+						return fmt.Errorf("`disk_size_in_gib` is required when`create_option` is `Empty`")
```
```suggestion
						return errors.New("`disk_size_in_gib` is required when`create_option` is `Empty`")
```

wyattfry commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1006 @@
+
+					lunExist := metadata.ResourceDiff.GetRawConfig().AsValueMap()["virtual_machine_profile"].AsValueSlice()[0].AsValueMap()["data_disk"].AsValueSlice()[0].AsValueMap()["lun"]
+					if lunExist.IsNull() {
+						return fmt.Errorf("`lun` is required when`create_option` is `Empty`")
```
```suggestion
						return errors.New("`lun` is required when`create_option` is `Empty`")
```

wyattfry commented on internal/services/computefleet/compute_fleet_resource.go:
```diff
@@ -0,0 +1,1006 @@
+
+			vmProfile := state.VirtualMachineProfile[0]
+			if vmProfile.SourceImageId != "" && len(vmProfile.SourceImageReference) > 0 {
+				return fmt.Errorf("only one of `source_image_id` and `source_image_reference` in `virtual_machine_profile` must be specified")
```
```suggestion
				return errors.New("only one of `source_image_id` and `source_image_reference` in `virtual_machine_profile` must be specified")
```

wyattfry reviewed (changes requested):
Hi @sinbai thanks for working on this eagerly awaited feature. I left some proposed changes for you.

sinbai committed:
fix comments
